framework/src/utils/KDTree.C

```diff
@@ -0,0 +1,68 @@
+/****************************************************************/
+/*               DO NOT MODIFY THIS HEADER                      */
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*           (c) 2010 Battelle Energy Alliance, LLC             */
+/*                   ALL RIGHTS RESERVED                        */
+/*                                                              */
+/*          Prepared by Battelle Energy Alliance, LLC           */
+/*            Under Contract No. DE-AC07-05ID14517              */
+/*            With the U. S. Department of Energy               */
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
+
+#include "KDTree.h"
+#include "MooseError.h"
+
+#include "libmesh/nanoflann.hpp"
+
+KDTree::KDTree(std::vector<Point> & master_points)
+  : _kd_tree(nullptr), _point_list_adaptor(master_points)
+{
+}
+
+void
+KDTree::buildTree()
+{
+  // Maximum number of points in each leaf of the Kd tree
+  unsigned int max_leaf_size = 10;
+  if (_kd_tree == nullptr)
+    _kd_tree.reset(new KdTreeT(LIBMESH_DIM,
+                               _point_list_adaptor,
+                               nanoflann::KDTreeSingleIndexAdaptorParams(max_leaf_size)));
+
+  mooseAssert(_kd_tree != nullptr, "KDTree was not properly initalized.");
+
+  _kd_tree->buildIndex();
+}
+
+void
+KDTree::neighborSearch(Point & query_point,
+                       unsigned int patch_size,
+                       std::vector<unsigned int> & return_index)
+{
+  // The query point has to be converted from a C++ array to a C array because nanoflann library
+  // expects C arrays.
+  const Real query_pt[] = {query_point(0), query_point(1), query_point(2)};
+
+  // Again size_t is used for ret_index because nanoflann expects that
```

Comment:
Thanks for the comment but this is actually a good thing. All indexing in C++ is std::size_t but people routinely (wrongly) use `unsigned int`. (Like every single loop counter everywhere, e.g. `unsigned int i = 0; i < v.size(); ++i`) Below you have another comment about turning std::size_t into an unsigned int with an extra vector. Why not just return and use the std::size_t everywhere? This is the default for all indexing operations. Even `vector::size()` returns a std::size_t but people store it into an unsigned int all the time. Yes everybody sucks at getting types right. 😉  Just use size_t here and don't worry about converting back and forth to unsigned int.